data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/S3Source.java

```diff
@@ -14,17 +14,20 @@
 import org.opensearch.dataprepper.model.event.Event;
 import org.opensearch.dataprepper.model.plugin.PluginFactory;
 import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.model.source.RequiresSourceCoordination;
 import org.opensearch.dataprepper.model.source.Source;
+import org.opensearch.dataprepper.model.source.SourceCoordinator;
 import org.opensearch.dataprepper.plugins.source.codec.Codec;
 import org.opensearch.dataprepper.plugins.source.ownership.BucketOwnerProvider;
 import org.opensearch.dataprepper.plugins.source.ownership.ConfigBucketOwnerProviderFactory;
 
 @DataPrepperPlugin(name = "s3", pluginType = Source.class, pluginConfigurationType = S3SourceConfig.class)
-public class S3Source implements Source<Record<Event>> {
+public class S3Source implements Source<Record<Event>>, RequiresSourceCoordination {
```

Comment:
I actually like the interface idea. But, seeing this here I realize that the source coordination is variably needed.

The S3 source does not require it for the SQS notification type. It only requires when it is configured to scan. Only in those cases do we want it.

Reply:
Yes it is not always needed. If no `source_coordination` config is specified in the `data-prepper-config.yaml`, then the `SourceCoordinator` class that is passed to the source will be null. The source can then decide whether it actually needs it or not.

Reply:
That makes sense. So if the source needs it and it is not available, the source will throw an exception.

I did just realize though - why not just make this injectable via the `@DataPrepperConstructor`? We won't even need the interface.